enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+
+Downgrade (all platforms):
+
+- `cluster-cloud-controller-manager-operator` de-provisions CCM `deployments` and stops working.
+
+The steps below apply to OpenStack only:
+
+Upgrade:
+
+- The `cluster-version-operator` starts `cluster-cloud-controller-manager-operator`.
+- `cluster-cloud-controller-manager-operator` creates all required resources for CCM (RBAC, Service Account, etc.).
+- `kube-apiserver` and `kube-controller-manager` are restarted without the `--cloud-provider` option.
+- `kubelet` is restarted with `--cloud-provider=external` option.
+
+Downgrade:
+
+- `cluster-cloud-controller-manager-operator` de-provisions CCM `deployments` and stops working.
+- `kube-apiserver` and `kube-controller-manager` are restarted with the `--cloud-provider=<some-platform>` option.
+- An old machine configuration is applied, which causes `kubelet` to restart with `--cloud-provider=<some-platform>` option.
+
+#### 4.7 -> 4.8
+
+Upgrade:
+
+- `cluster-cloud-controller-manager-operator` creates resources for any in-tree provider defaulting to this in 4.8 (expecting: AWS, Azure, GCP, vSphere).
+- `kube-apiserver` and `kube-controller-manager` are restarted without the `--cloud-provider` option.
+- `kubelet` is restarted with `--cloud-provider=external` option.
+
+Downgrade:
+
+- `cluster-cloud-controller-manager-operator` de-provisions CCM resources for any platform, except OpenStack.
+- `kube-apiserver` and `kube-controller-manager` are restarted with the `--cloud-provider=<some-platform>` option.
+- An old machine configuration is applied, which causes `kubelet` to restart with `--cloud-provider=<some-platform>` option (excluding OpenStack).
+
+#### 4.8 and later
+
+Upgrade:
+
+- `cluster-cloud-controller-manager-operator` updates the state of resources for any in-tree provider defaulting to this since previous releases.
+
+Downgrade:
+
+- `cluster-cloud-controller-manager-operator` downgrades the state of resources for any in-tree provider defaulting to this since previous releases.
+
+### Version Skew Strategy
+
+See the upgrade/downgrade strategy.
+
+### Action plan (for OpenStack)
+
+#### Build OpenStack CCM image by OpenShift automation
+
+To start using OpenStack CCM in OpenShift we need to build its image and make sure it is a part of the OpenShift release image. The CCM image should be automatically tested before it becomes available.
+The upstream repo provides the Dockerfile, so we can reuse it to complete the task. Upstream image is already available in [Dockerhub](https://hub.docker.com/r/k8scloudprovider/openstack-cloud-controller-manager).
+
+Actions:
+
+- Configure CI operator to build OpenStack CCM image.
+
+CI operator will run containerized and End-to-End tests and also push the resulting image in the OpenShift Quay account.
+
+#### Test Cloud Controller Manager manually
+
+When all required components are built, we can manually deploy OpenStack CCM and test how it works.
```

Comment:
Manual is ok for dev testing, we need a proper set of automatic tests ensuring full functionality of the operator and its operands.

Reply:
agree! now we don't have a dedicated openstack-upgrade ci job, but it would be nice to have one to reduce migration risks and minimize the number of potential errors.